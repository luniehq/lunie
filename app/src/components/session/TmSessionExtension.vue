<template>
  <SessionFrame :icon="`laptop`">
    <div class="session-container">
      <h2 class="session-title">Use Lunie Browser Extension</h2>
      <div v-if="!extension.enabled" class="session-main">
        <p>
          Looks like you don't have the Lunie browser extension installed yet.
          Head over to the
          <a
            href="http://bit.ly/lunie-ext"
            target="_blank"
            rel="noopener norefferer"
            >Chrome Web Store</a
          >
          to quickly install the extension.
        </p>
      </div>

      <div v-else-if="accounts.length" class="session-main">
        <p class="extension-message">
          Below is a list of accounts we've received from the Lunie browser
          extension.
        </p>
        <AccountList
          :accounts="accounts"
          :button-action="signInAndRedirect"
          :button-text="`Use Account`"
        />
      </div>

      <div v-else class="session-main">
        <p class="extension-message">
          Looks like you don't have any addresses in the Lunie extension yet.
          Click on the extension icon in your browser and add an address now.
        </p>
      </div>
    </div>
  </SessionFrame>
</template>

<script>
import AccountList from "account/AccountList"
import SessionFrame from "common/SessionFrame"
import { mapState, mapGetters } from "vuex"
export default {
  name: `session-extension`,
  components: {
    AccountList,
    SessionFrame,
  },
  data: () => ({
    connectionError: null,
    address: null,
  }),
  computed: {
    ...mapState([`extension`]),
    ...mapGetters([`networks`]),
    accounts() {
      return this.extension.accounts
    },
  },
  mounted() {
    this.getAddressesFromExtension()
  },
  methods: {
    getAddressesFromExtension() {
      this.$store.dispatch("getAddressesFromExtension")
    },
    async signIn(account) {
      this.$store.dispatch(`signIn`, {
        sessionType: `extension`,
        address: account.address,
        networkId: account.network ? account.network : "cosmos-hub-mainnet", // defaulting to cosmos-hub-mainnet
      })
    },
    async signInAndRedirect(account) {
      const accountNetwork = await this.$store.dispatch("getNetworkByAccount", {
        account,
      })
      await this.signIn(account)
      this.$router.push({
        name: "portfolio",
        params: {
          networkId: accountNetwork.slug,
        },
      })
    },
  },
}
</script>
