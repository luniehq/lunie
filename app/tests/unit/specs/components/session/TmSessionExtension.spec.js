import { shallowMount } from "@vue/test-utils"
import TmSessionExtension from "session/TmSessionExtension"

describe(`TmSessionExtension`, () => {
  let wrapper, $store, $router

  const accounts = [
    {
      address: "cosmos1234",
      network: "cosmos-hub-mainnet",
      slug: "cosmos-hub",
    },
    {
      address: "cosmos15678",
      network: "gaia-testnet",
      slug: "cosmos-hub-testnet",
    },
  ]

  const networks = [
    {
      id: "gaia-testnet",
      chain_id: "gaia-123",
      logo_url: "cosmos-logo.png",
      testnet: true,
      title: "Cosmos Hub Test",
      slug: "gaia",
    },
    {
      id: "cosmos-hub-mainnet",
      chain_id: "cosmoshub",
      logo_url: "cosmos-logo.png",
      testnet: false,
      title: "Cosmos Hub",
      slug: "cosmos-hub",
    },
  ]

  beforeEach(() => {
    const state = {
      extension: {
        enabled: true,
        accounts,
      },
    }

    $store = {
      getters: {
        networkSlug: "cosmos-hub",
        networks,
      },
      commit: jest.fn(),
      dispatch: jest.fn(),
      state,
    }
    $router = {
      push: jest.fn(),
    }
    wrapper = shallowMount(TmSessionExtension, {
      mocks: {
        $store,
        $router,
      },
    })
  })

  it("should sign in with the selected account", async () => {
    $store.dispatch = jest.fn(() => Promise.resolve(networks[1]))
    await wrapper.vm.signInAndRedirect({
      network: "cosmos-hub-mainnet",
      address: "cosmos1",
    })

    expect($router.push).toHaveBeenCalledWith({
      name: "portfolio",
      params: { networkId: "cosmos-hub" },
    })
    expect($store.dispatch).toHaveBeenCalledWith("signIn", {
      sessionType: `extension`,
      address: "cosmos1",
      networkId: "cosmos-hub-mainnet",
    })
  })

  it(`should show a list of all accounts in the extension`, () => {
    expect(wrapper.element).toMatchSnapshot()
  })

  it(`should show a guide to install the extension if no extension is installed`, () => {
    wrapper.vm.extension.enabled = false
    expect(wrapper.element).toMatchSnapshot()
  })

  it("should load addresses on mount", () => {
    expect($store.dispatch).toHaveBeenCalledWith("getAddressesFromExtension")
  })
})
