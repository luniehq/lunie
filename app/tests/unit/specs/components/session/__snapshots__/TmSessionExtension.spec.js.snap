// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TmSessionExtension should show a guide to install the extension if no extension is installed 1`] = `
<sessionframe-stub
  icon="laptop"
>
  <div
    class="session-container"
  >
    <h2
      class="session-title"
    >
      Use Lunie Browser Extension
    </h2>
     
    <div
      class="session-main"
    >
      <p>
        
        Looks like you don't have the Lunie browser extension installed yet.
        Head over to the
        
        <a
          href="http://bit.ly/lunie-ext"
          rel="noopener norefferer"
          target="_blank"
        >
          Chrome Web Store
        </a>
        
        to quickly install the extension.
      
      </p>
    </div>
  </div>
</sessionframe-stub>
`;

exports[`TmSessionExtension should show a list of all accounts in the extension 1`] = `
<sessionframe-stub
  icon="laptop"
>
  <div
    class="session-container"
  >
    <h2
      class="session-title"
    >
      Use Lunie Browser Extension
    </h2>
     
    <div
      class="session-main"
    >
      <p
        class="extension-message"
      >
        
        Below is a list of accounts we've received from the Lunie browser
        extension.
      
      </p>
       
      <accountlist-stub
        accounts="[object Object],[object Object]"
        buttonaction="function () { [native code] }"
        buttontext="Use Account"
      />
    </div>
  </div>
</sessionframe-stub>
`;
